### PR TITLE
feat: ENT-5364 Added pathway indexing for Algolia

### PR DESCRIPTION
**Pull request by irfanuddinahmad**
## Description

This PR adds Pathways type content metadata to the Algolia search index
image

## Ticket Link

[ENT-5364](https://openedx.atlassian.net/browse/ENT-5364)

### CODE REVIEW

File: enterprise_catalog/apps/api/tasks.py [hunk 1]
```diff
@@ -462,7 +463,10 @@ def index_enterprise_catalog_in_algolia_task(self, force=False):  # pylint: disa
         indexable_program_keys, nonindexable_program_keys = partition_program_keys_for_indexing(
             programs_content_metadata,
         )
-        indexable_content_keys = indexable_course_keys + indexable_program_keys
+        pathways_content_metadata = ContentMetadata.objects.filter(content_type=PATHWAY)
+        pathways_content_keys = set(pathway_metadata.content_key for pathway_metadata in pathways_content_metadata)
+        indexable_pathways_keys = list(pathways_content_keys)
+        indexable_content_keys = indexable_course_keys + indexable_program_keys + indexable_pathways_keys
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/api/tasks.py]: Can we do it in 1 line?
```
indexable_pathways_keys = ContentMetadata.objects.filter(content_type=PATHWAY). distinct().values_list('content_key', flat=True)
```

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 2]
```diff
@@ -337,6 +338,100 @@ def _get_course_program_field(course, field):
...
+
+    Returns:
+        list: a union of pathway programs and courses availability.
+    """
+    availability = set()
+    steps = pathway.get('steps') or []
+    for step in steps:
+        courses = step.get('courses') or []
+        for course in courses:
+            course_metadata = ContentMetadata.objects.filter(content_key=course.get('key')).first()
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py]: can we use `get_pathway_course_keys ` and `get_pathway_program_uuids ` utilities and query the database only once to get all the records?

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 3]
```diff
@@ -337,6 +338,100 @@ def _get_course_program_field(course, field):
...
+        programs = step.get('programs') or []
+        for program in programs:
+            program_metadata = ContentMetadata.objects.filter(content_key=program.get('uuid')).first()
+            if program_metadata:
+                program_status = get_program_availability(program_metadata.json_metadata)
+                availability.update(program_status)
+    return list(availability)
+
+
+def get_pathway_skill_names(pathway):
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py]: I have added skills in search/all so we don't need to get skills again here.

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 4]
```diff
@@ -829,6 +924,13 @@ def _algolia_object_from_product(product, algolia_fields):
             'prices': get_program_prices(searchable_product),
             'banner_image_url': get_program_banner_image_url(searchable_product),
         })
+    elif searchable_product.get('content_type') == PATHWAY:
+        searchable_product.update({
+            'course_keys': get_pathway_course_keys(searchable_product),
+            'programs': get_pathway_program_uuids(searchable_product),
+            'availability': get_pathway_availability(searchable_product),
+            'skill_names': get_pathway_skill_names(searchable_product),
+        })
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py:1005]: I think we also add information like, `language`, `partners `, `subjects `, `level_type ` `banner_image`

**Review by HammadAhmadWaqas — CHANGES_REQUESTED** [review]

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 2 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py]: Can you elaborate?

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 4 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py:1005]: how are these defined for a pathway?

Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py:1005]: Do we need the banner in the index?

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 2 above)
Comment (HammadAhmadWaqas) [reply enterprise_catalog/apps/catalog/algolia_utils.py]: Right now you are making database queries in a loop. We can query only once and process all those content metadata.

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 4 above)
Comment (HammadAhmadWaqas) [reply enterprise_catalog/apps/catalog/algolia_utils.py:1005]: The same way we are getting them for programs. And yes we need that image on the Pathways card.

**Review by HammadAhmadWaqas — COMMENTED** [review]
Queries Answered

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 4 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py:1005]: I think we need to define the level_type and language for a pathway ...

Comment (HammadAhmadWaqas) [reply enterprise_catalog/apps/catalog/algolia_utils.py:1005]: I think we can get it same way we are getting in Program

**Review by HammadAhmadWaqas — APPROVED** [review]

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 5]
```diff
@@ -339,6 +339,139 @@ def _get_course_program_field(course, field):
...
+        str: url to large size image
+    """
+    images = pathway.get('banner_image', {})
+    try:
+        return images.get('large').get('url')
+    except (KeyError, AttributeError):
+        return None
+
+
+def get_pathway_partners(pathway):
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py:421]: Wo have partner field in pathways own object.

File: enterprise_catalog/apps/catalog/tests/test_algolia_utils.py [hunk 6]
```diff
@@ -874,3 +880,299 @@ def test_get_program_level_type(self, program_metadata, course_metadata, expecte
...
+        """
+        Assert that the programs associated with a pathway are properly parsed.
+        """
+        # for i in range(2):
+        #     ContentMetadataFactory.create(
+        #         content_key=pathway_metadata['steps'][i]['courses'][i]['key'],
+        #         content_type=COURSE,
+        #         json_metadata=course_metadata[i],
+        #     )
+
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/tests/test_algolia_utils.py:942]: Do we need to remove this commented code?

**Pull request merged**